src/GetAssemblyInfoViz.cpp

```diff
@@ -0,0 +1,196 @@
+        int line_number = 0;
+        while (getline(assembly_file, assembly_line)) {
+            line_number++;
+            assembly_HTML << assembly_line << "\n";
+            add_line_number(assembly_line, line_number);
+        }
+        assembly_file.close();
+    }
+
+    assembly_HTML << "</pre>\n";
+    assembly_HTML << "</div>\n";
+}
+
+string GetAssemblyInfoViz::get_assembly_filename(const string &filename) {
+    string assembly_filename = "./" + filename;
+    assembly_filename.replace(assembly_filename.find(".stmt.viz.html"), 15, ".s");
+    return assembly_filename;
+}
+
+void GetAssemblyInfoViz::add_line_number(string &assembly_line, int line_number) {
+    for (auto &marker : for_loop_markers) {
+        add_line_number_for_loop(assembly_line, marker, line_number);
+    }
+    for (auto &marker : producer_consumer_markers) {
+        add_line_number_prod_cons(assembly_line, marker, line_number);
+    }
+}
+
+void GetAssemblyInfoViz::add_line_number_for_loop(string &assembly_line,
+                                                  AssemblyInfoForLoop &marker, int line_number) {
+    // start of for loop
+    if (std::regex_search(assembly_line, marker.regex_start)) {
+
+        // check if marker is already present
+        auto it = node_to_line_numbers_for_loops.find(marker.node);
+        if (it == node_to_line_numbers_for_loops.end()) {
+            ForLoopLineNumber for_loop_line_number;
+            for_loop_line_number.start_line = line_number;
+            for_loop_line_number.end_line = -1;
+            node_to_line_numbers_for_loops[marker.node] = for_loop_line_number;
+        } else {
+            it->second.start_line = line_number;
+        }
+    }
+
+    // end of for loop
+    if (std::regex_search(assembly_line, marker.regex_end)) {
+
+        // check if marker is already present
+        auto it = node_to_line_numbers_for_loops.find(marker.node);
+        if (it == node_to_line_numbers_for_loops.end()) {
+            ForLoopLineNumber for_loop_line_number;
+            for_loop_line_number.end_line = line_number;
+            for_loop_line_number.start_line = -1;
+            node_to_line_numbers_for_loops[marker.node] = for_loop_line_number;
+        } else {
+            it->second.end_line = line_number;
+        }
+    }
+}
+void GetAssemblyInfoViz::add_line_number_prod_cons(string &assembly_line,
+                                                   AssemblyInfoProdCons &marker, int line_number) {
+    if (std::regex_search(assembly_line, marker.regex)) {
+        node_to_line_number_prod_cons[marker.node] = line_number;
+    }
+}
+
+void GetAssemblyInfoViz::visit(const ProducerConsumer *op) {
+    producer_consumer_count++;
+
+    string assembly_marker = "%\"";
+    assembly_marker += to_string(producer_consumer_count);
+    assembly_marker += op->is_producer ? "_produce " : "_consume ";
+    assembly_marker += op->name;
+
+    // replace all $ with \$
+    std::regex dollar("\\$");
```

Comment:
Do you need to use `regex` for this? Seems like the `replace_all()` function in Util.h would probably do the same thing a bit more cheaply